query/executor_test.go

```diff
@@ -595,6 +604,313 @@ func TestQueryExecutor_InvalidSource(t *testing.T) {
+	require.NoError(t, err, "parse query")
+
+	e := NewQueryExecutor()
+	mockWatcher := newMockWatcher(t, "query.log")
+	defer mockWatcher.Close()
+	e.WithLogWriter(mockWatcher, context.Background())
+
+	e.StatementExecutor = &StatementExecutor{
+		ExecuteStatementFn: func(stmt influxql.Statement, ctx *query.ExecutionContext) error {
+			require.Equal(t, uint64(1), ctx.QueryID, "query ID")
+			return nil
+		},
+	}
+
+	discardOutput(e.ExecuteQuery(q, query.ExecutionOptions{}, nil))
+
+	deadline := time.Now().Add(time.Second)
+	for time.Now().Before(deadline) {
+		if mockWatcher.ChangeEvents.Load() == 1 {
+			break
+		}
+
+		runtime.Gosched()
+	}
+
+	require.True(t, mockWatcher.Contains(`SELECT count(value) FROM cpu`))
+	require.Equal(t, uint64(0), mockWatcher.RemovedPaths.Load())
+	require.Equal(t, uint64(1), mockWatcher.AddedPaths.Load())
+	require.Equal(t, uint64(0), mockWatcher.ChangeEvents.Load())
+
+	mockWatcher.MockEvents <- fsnotify.Event{
+		Name: "Rename",
+		Op:   fsnotify.Rename,
+	}
+
+	q, err = influxql.ParseQuery(`SELECT count(value) FROM mem`)
+
+	e.StatementExecutor = &StatementExecutor{
+		ExecuteStatementFn: func(stmt influxql.Statement, ctx *query.ExecutionContext) error {
+			require.Equal(t, uint64(2), ctx.QueryID, "query ID")
+			return nil
+		},
+	}
+
+	discardOutput(e.ExecuteQuery(q, query.ExecutionOptions{}, nil))
+
+	for time.Now().Before(deadline) {
+		if mockWatcher.ChangeEvents.Load() == 1 {
+			break
+		}
+
+		runtime.Gosched()
+	}
+
+	require.True(t, mockWatcher.Contains(`SELECT count(value) FROM mem`))
+
+	require.Equal(t, uint64(1), mockWatcher.ChangeEvents.Load())
+	require.Equal(t, uint64(1), mockWatcher.RemovedPaths.Load())
+
+	err = mockWatcher.Close()
+	require.NoError(t, err)
+}
+
+func TestQueryExecutor_RemoveEvent(t *testing.T) {
+	q, err := influxql.ParseQuery(`SELECT count(value) FROM cpu`)
+	require.NoError(t, err, "parse query")
+
+	e := NewQueryExecutor()
+	mockWatcher := newMockWatcher(t, "query.log")
+	defer mockWatcher.Close()
+	e.WithLogWriter(mockWatcher, context.Background())
+
+	e.StatementExecutor = &StatementExecutor{
+		ExecuteStatementFn: func(stmt influxql.Statement, ctx *query.ExecutionContext) error {
+			require.Equal(t, uint64(1), ctx.QueryID, "query ID")
+			return nil
+		},
+	}
+
+	discardOutput(e.ExecuteQuery(q, query.ExecutionOptions{}, nil))
+
+	deadline := time.Now().Add(time.Second)
+	for time.Now().Before(deadline) {
+		if mockWatcher.AddedPaths.Load() == 1 {
+			break
+		}
+
+		runtime.Gosched()
+	}
+
+	require.True(t, mockWatcher.Contains(`SELECT count(value) FROM cpu`))
+	require.Equal(t, uint64(0), mockWatcher.RemovedPaths.Load())
+	require.Equal(t, uint64(1), mockWatcher.AddedPaths.Load())
+	require.Equal(t, uint64(0), mockWatcher.ChangeEvents.Load())
+
+	mockWatcher.MockEvents <- fsnotify.Event{
+		Name: "Remove",
+		Op:   fsnotify.Remove,
+	}
+
+	for time.Now().Before(deadline) {
+		if mockWatcher.ChangeEvents.Load() == 1 {
+			break
```

Comment:
Set a bool variable here to see if this is the loop exit or the `ChangeEvents.Load` is

Reply:
👍